ead/edhoc-ead-zeroconf/src/lib.rs

```diff
@@ -115,28 +327,636 @@ pub fn r_process_ead_1(ead_1: EADItem) -> Result<(), ()> {
     Ok(())
 }
 
-pub fn r_prepare_ead_2() -> Option<EADItem> {
-    let mut ead_2 = EADItem::new();
+pub fn r_prepare_ead_2(voucher_response: &Option<EdhocMessageBuffer>) -> Option<EADItem> {
+    let mut output: Option<EADItem> = None;
 
-    // add the label to the buffer (non-critical)
-    ead_2.label = EAD_ZEROCONF_LABEL;
-    ead_2.is_critical = true;
+    if let Some(voucher_response) = voucher_response {
+        let mut ead_2 = EADItem::new();
 
-    // TODO: append Voucher (H(message_1), CRED_V) to the buffer
+        // FIXME: we probably don't want to parse the voucher response here, but rather receive only the 'voucher' part, already parsed
+        let (_message_1, voucher, _opaque_state) =
+            parse_voucher_response(voucher_response).unwrap();
+
+        output = Some(EADItem {
+            label: EAD_ZEROCONF_LABEL,
+            is_critical: true,
+            value: Some(voucher),
+        });
+    }
 
     // NOTE: see the note in lib.rs::test_ead
-    // state.protocol_state = EADResponderProtocolState::WaitMessage3;
+    // set as completed even if the voucher response is not present
     ead_responder_set_global_state(EADResponderState {
         protocol_state: EADResponderProtocolState::Completed,
     });
 
-    Some(ead_2)
+    output
 }
 
-pub fn r_process_ead_3(ead_3: EADItem) -> Result<(), ()> {
+pub fn r_process_ead_3(_ead_3: EADItem) -> Result<(), ()> {
     // TODO: maybe retrive CRED_U from a Credential Database
 
     // state.protocol_state = EADResponderProtocolState::Completed;
 
     Ok(())
 }
+
+fn parse_voucher_response(
+    voucher_response: &EdhocMessageBuffer,
+) -> Result<
+    (
+        EdhocMessageBuffer,
+        EdhocMessageBuffer,
+        Option<EdhocMessageBuffer>,
+    ),
+    (),
+> {
+    let mut message_1 = EdhocMessageBuffer::new();
+    let mut voucher = EdhocMessageBuffer::new();
```

Comment:
I changed this for consistency, but note that since we need to put the voucher in an EADItem struct, we sill need to allocate one large buffer since the `EADItem.value` is an `EdhocMessageBuffer`.